Dynamo tensor aliasing guards, dedup graphargs

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103205
* __->__ #104921

The story here is relatively simple - when we go to wrap a tensor, we (1) ensure that it is a real, not fake tensor (2) check if we have seen it before. (3) If we have seen it, we create a positive alias guard and return the associated variable. If not, we proceed. 

By short circuiting here, we avoid lifting it to a graph input, and guarantee that the only names passed to tensors are unique. This allows us to guard on the unique relationships (pyboject addresses, aka IDs, cannot match) to give us guards for negative aliases. 

cc @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

voznesenskym committed:
Dynamo general purpose tensor aliasing

Tests

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo general purpose tensor aliasing"


Tests

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo general purpose tensor aliasing"


Tests

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo general purpose tensor aliasing"


The story here is relatively simple - when we go to wrap a tensor, we (1) ensure that it is a real, not fake tensor (2) check if we have seen it before. (3) If we have seen it, we create a positive alias guard and return the associated variable. If not, we proceed. 

By short circuiting here, we avoid lifting it to a graph input, and guarantee that the only names passed to tensors are unique. This allows us to guard on the unique relationships (pyboject addresses, aka IDs, cannot match) to give us guards for negative aliases. 

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo general purpose tensor aliasing"


The story here is relatively simple - when we go to wrap a tensor, we (1) ensure that it is a real, not fake tensor (2) check if we have seen it before. (3) If we have seen it, we create a positive alias guard and return the associated variable. If not, we proceed. 

By short circuiting here, we avoid lifting it to a graph input, and guarantee that the only names passed to tensors are unique. This allows us to guard on the unique relationships (pyboject addresses, aka IDs, cannot match) to give us guards for negative aliases. 

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo general purpose tensor aliasing"


The story here is relatively simple - when we go to wrap a tensor, we (1) ensure that it is a real, not fake tensor (2) check if we have seen it before. (3) If we have seen it, we create a positive alias guard and return the associated variable. If not, we proceed. 

By short circuiting here, we avoid lifting it to a graph input, and guarantee that the only names passed to tensors are unique. This allows us to guard on the unique relationships (pyboject addresses, aka IDs, cannot match) to give us guards for negative aliases. 

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -899,6 +899,21 @@ def wrap_tensor(self, value: torch.Tensor):
+        if not self.tx.output.export:
+            # Export has (supposedly) valid cases for fake tensors as inputs here.
+            # I am not convinced, atm, but out of scope for what this assert was added for (protecting value checks
+            # in real_value_tensor_positive_aliases in the common case)
```
The PR in question is https://github.com/pytorch/pytorch/pull/100017 I recommend reading through it. I will say that wanting to do a PT2 export without actually instantiating the model with real parameters seems like a legitimate use case to me.

voznesenskym replied:
Thats fair, but for that case I suppose we don't care about these guards as much, and can probably just not have the dedup and aliasing guards for export...

ezyang commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -292,12 +294,25 @@ PyObject* TensorGuards_check(TensorGuards* self, PyObject* args) {
+  // create guards for negative alias (X is not Y) as that is an N^2
+  // relationship. Instead, we rely on the uniqueness upstream to verify, at
+  // check_fn time (this function).
+  std::set<PyObject*> unique_tensors;
```
Consider using flat_hash_map instead, std::set is famously slow

voznesenskym replied:
Sounds good!

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -899,6 +899,21 @@ def wrap_tensor(self, value: torch.Tensor):
+        if value in self.tx.output.real_value_tensor_positive_aliases:
+            stored_value = self.tx.output.real_value_tensor_positive_aliases[value]
+            # TODO(voz): Decently common pattern, refactor at some point.
+            dup_guard = self._make_dupe_guard(stored_value)
```
This function has a TODO for handling globals. You are making this ping a lot more frequently, it seems like that should get fixed now?

voznesenskym replied:
The TODO is a little more nuanced. It is not for globals, but for dupe guards that cross alias across a local and global. 

It's also overly restrictive, let me see if I can alleviate it a little.

ezyang commented:
I triggered a perf run at https://github.com/pytorch/pytorch/actions/runs/5520442947

ezyang commented on test/dynamo/test_recompiles.py:
```diff
@@ -151,3 +151,25 @@ def run_with_automatic():
         self.assertEqual(with_automatic.frame_count, 3)
         self.assertEqual(with_automatic.op_count, 3)
+
+    def test_aliasing_guard_failures(self):
```
By the way, I would not call this aliasing. Typically, when I say two tensors alias each other, their data pointers overlap. However, typically, these would be distinct tensor PyObjects!

voznesenskym replied:
@jansel and I have been discussing this offline as "aliasing", and the point here is to protect cases where `c` is an alias of `a` because the user passed in `x` to both of them. 

The terms we've been using are positive alias guards (x is y) and negative alias guards (x is not y) - open to bikeshedding the names, but this does feel apt here.

ezyang replied:
I'd vote for "Python object identity" which is clear and unambiguous with data aliasing.

ezyang commented:
This seems basically OK, if a bit risky. (But I guess not that risky, since CI is all green.)

If you are enforcing PyObject uniqueness for tensor inputs, do you not also have to do this for all other objects that we may guard on (e.g., lists)? It would be good for @jansel to take a look too.

bdhirsh commented on test/dynamo/test_recompiles.py:
```diff
@@ -151,3 +151,25 @@ def run_with_automatic():
+
+        foo(x, y, x)
+        # Recompile, alias changed
+        self.assertEqual(cnt.frame_count, 2)
```
Basic question: what's the relationship dynamo's deduping in this PR vs. AOTAutograd's de-duping logic? (Is this PR meant to replace the version of deduping in AOTAutograd?) https://github.com/pytorch/pytorch/blob/999abd56a7be3d73f0d2e9ab06356c7281a3307d/torch/_functorch/aot_autograd.py#L2040

voznesenskym replied:
This does not cover params - only input tensors guarded through `___check_tensors`

voznesenskym replied:
Dynamo should be correct standalone, without aotautograd, and AOTAutograd does not guard negative (is not) aliases.

There is a lot of overlap here, and we should probably dedup at some point.

voznesenskym replied:
@bdhirsh mind if I close this out?

bdhirsh replied:
yep for sure :) thanks for the context

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -292,12 +294,25 @@ PyObject* TensorGuards_check(TensorGuards* self, PyObject* args) {
     if (Py_TYPE(item) != checks[i].pytype) {
       Py_RETURN_FALSE;
     }
+    auto it = unique_tensors.find(item);
```
Good call

voznesenskym commented:
> This seems basically OK, if a bit risky. (But I guess not that risky, since CI is all green.)
> 
> If you are enforcing PyObject uniqueness for tensor inputs, do you not also have to do this for all other objects that we may guard on (e.g., lists)? It would be good for @jansel to take a look too.

I am only enforcing uniqueness for tensor inputs because I want tensor aliasing. We could do this across the board, but I only care about it for tensors.

jansel commented on test/dynamo/test_recompiles.py:
```diff
@@ -151,3 +151,25 @@ def run_with_automatic():
+        # No recompile, alias preserved
+        self.assertEqual(cnt.frame_count, 1)
+
+        foo(x, y, x)
```
We should check correctness in these tests as well.

voznesenskym replied:
added

voznesenskym committed:
Update on "Dynamo tensor aliasing guards, dedup graphargs"


The story here is relatively simple - when we go to wrap a tensor, we (1) ensure that it is a real, not fake tensor (2) check if we have seen it before. (3) If we have seen it, we create a positive alias guard and return the associated variable. If not, we proceed. 

By short circuiting here, we avoid lifting it to a graph input, and guarantee that the only names passed to tensors are unique. This allows us to guard on the unique relationships (pyboject addresses, aka IDs, cannot match) to give us guards for negative aliases. 

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo tensor aliasing guards, dedup graphargs"


The story here is relatively simple - when we go to wrap a tensor, we (1) ensure that it is a real, not fake tensor (2) check if we have seen it before. (3) If we have seen it, we create a positive alias guard and return the associated variable. If not, we proceed. 

By short circuiting here, we avoid lifting it to a graph input, and guarantee that the only names passed to tensors are unique. This allows us to guard on the unique relationships (pyboject addresses, aka IDs, cannot match) to give us guards for negative aliases. 

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym committed:
Update on "Dynamo tensor aliasing guards, dedup graphargs"


The story here is relatively simple - when we go to wrap a tensor, we (1) ensure that it is a real, not fake tensor (2) check if we have seen it before. (3) If we have seen it, we create a positive alias guard and return the associated variable. If not, we proceed. 

By short circuiting here, we avoid lifting it to a graph input, and guarantee that the only names passed to tensors are unique. This allows us to guard on the unique relationships (pyboject addresses, aka IDs, cannot match) to give us guards for negative aliases. 

cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

voznesenskym commented:
<img width="1704" alt="image" src="https://github.com/pytorch/pytorch/assets/4755252/a3889878-1499-491d-9484-a66ab4ed7b4d">

Modest (1-3%) to neutral speedups, as expected

voznesenskym commented:
Look, pa, no failures

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
https://github.com/pytorch/pytorch/pull/103205/files#diff-9f0663783bcd93e948e0491ef61b48123bdc9977bcc632fd707da578df13bfa1R879-R891 

Mistaken merge there - this should be updated now.

lezcano commented on torch/_dynamo/output_graph.py:
```diff
@@ -232,6 +232,11 @@ def __init__(
+
+        # Used to maintain an alias between real values variable tracker for tensors we have seen.
+        # This map ensures that the only tensors in graph inputs, and the only tensors in guards are unique.
+        self.real_value_tensor_positive_aliases = WeakTensorKeyDictionary()
```
@voznesenskym where is this map populated?
At the moment, in master, we are checking whether something is in this variable before using it, so is it being used at all?
